README.md
# Linkhandler

This is a fork of the linkhandler TYPO3 extension.

## About this fork

This is an officious version of extension linkhandler. It is the result of my own
cleanups and of other forks, merged into this one.

The master version was duly tested with TYPO3 CMS 7.2. It seems to work fine with
up to 7.4, but is not fully tested.

For a version compatible with TYPO3 CMS 6.2, please use the TYPO3_6-2 branch.

## Additional features

This Extension offers some additional features compared to the original version.

### Multiple configurations for the same table

It is now possible to create links to the same table within different configurations. E.g. you could link
to different content types to different target pages.

For this to work the links now consists of four parts instead of three. The old link block looked like this:

```
record:<table_name>:<uid>
```

The new link block looks like this:

```
record:<config_key>:<table_name>:<uid>
```

This feature is backward compatible. If an old link is detected that consists only of three parts the first
configuration found for the linked table will be used.

### Record filtering

The records that are displayed in the list can be filtered by SQL queries. You can define a search query
for each table configured in ```listTables```. This is an example configuration for using the
```additionalSearchQueries``` option in the TSConfig:

```
mod.tx_linkhandler.tx_myext_imagelinks {
	label = Image link
	listTables = tt_content
	additionalSearchQueries {
		tt_content = AND tt_content.CType='image'
	}
}
```

### Page tree mount points

You can configure mount points for the page tree that is displayed in the element browser.

For example if you only want to diplay the pages where your news records are located
(in this example PID 123 and 234) you can use the following Page TSConfig:

```
mod.tx_linkhandler.tx_news_news.pageTreeMountPoints {
	1 = 123
	2 = 234
}
RTE.default.tx_linkhandler.tx_news_news.pageTreeMountPoints < mod.tx_linkhandler.tx_news_news.pageTreeMountPoints
```

### Linkvalidator support

This linkhandler version comes with its own linkvalidator link type that supports the new link format with four parameters
and provides some additional features that are not merged yet to the core.

To use it you have to adjust your linkvalidator TSConfig:

```
mod.linkvalidator {
	linktypes = db,external,tx_linkhandler
}
```

Please not that you need to use ```tx_linkhandler``` instead of ```linkhandler``` which is the default link type that comes with the core.

This link type comes with an additional configuration option that allows the reporting of links that point to  hidden records:

```
mod.linkvalidator {
	tx_linkhandler.reportHiddenRecords = 1
}
```

For this additional option to work this pending TYPO3 patch is required: https://review.typo3.org/#/c/26499/ (Provide TSConfig to link checkers).
There is a TYPO3 6.2 fork that already implements the required patches (and some more) at Github: https://github.com/Intera/TYPO3.CMS

### Pass on parent typolink configuration

*This feature is experimental and might change in the future!* Testing and suggestions welcome :)

A config option called ```overrideParentTypolinkConfiguration``` is available. When this option is enabled for a tab configuration,
the original configuration passed to the ```typolink()``` method will be used as the basis for the final typolink configuration that
is used in linkhandler. Here is an example:

Imagine you have a typolink like this:

```
lib.myobj = TEXT
lib.myobj {
	value = Hello World
	typolink.parameter = record:tx_news_news:tx_news_domain_model_news:1
	typolink.no_cache = 1
}
```

When you now configure for example the news records like this, the ```no_cache``` option from the ```lib.myobj.typolink``` block
will be passed on to the typolink call used by linkhandler:

```
plugin.tx_linkhandler.tx_news_news {
	overrideParentTypolinkConfiguration = 1
}
```

You can still override this behavior in the linkhandler configuration though because the ```typolink`` block is merged into
the parent configuration:

```
plugin.tx_linkhandler.tx_news_news {
	overrideParentTypolinkConfiguration = 1
	typolink.no_cache = 0
}
```

The only value that is **always** overwritten is the ```parameter``` configuration.


### Additional goodies

* When editing a link the correct tab will open automatically.
* The searchbox below the record list can be disabled by setting ```enableSearchBox = 0``` in the tab configuration in TSConfig.
* SoftReference handling using signal slots, TYPO3 patch pending: https://review.typo3.org/27746/
* The current link is displayed in a nice label consisting of the localized table label and the linked record title.

## Tips & Tricks

### Link browser width

You can use TSConfig to increase the with of the link browser windows in the Backend to prevent problem with the styles:

```
RTE {
	default {
		buttons {
			link {
				dialogueWindow {
					width = 600
				}
			}
		}
	}
}
```

## Missing Feature

The last missing feature in this version is the handling of the "Save and show" button.

>>>>>>> intera6-2


ChangeLog
2015-11-20 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Refactored for new LinkHandler API, requires TYPO3 7 LTS

2013-06-11 Alexander Stehlik <astehlik@intera.de>
	* Fully refactored version, compatible with TYPO3 6.2
	* Warning! Some new features like softrefs and "Save and show" are still missing

2013-04-23 Bernhard Kraft  <kraft@webconsulting.at>
	* Fixed bug #47505: double ampersands in link-browse popup

2013-03-12 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Heavily patched for TYPO3 6.0 compatibility.
	* Commented out all code in ext_localconf.php related to older versions of TYPO3. This is better for performance.

09-11-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Remove URL verification, just force "header" redirect because of problems if the requested pages is protected by HTTP authentication.

09-11-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Keep attention on the languagevisibility XCLASS
	* Change default value for "options.saveDocView.tt_news"

03-11-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Make the t3lib_softrefproc XCLASS optional during the extension manager.
	* Fixed bug #5199: Replace "t3lib_div::makeInstanceClassName()"

02-11-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Add feature #5191: Provide XCLASS to enable bug fix #10827: Hide "Save and View"-button when editing a content-element
	* Add some "conflicts" constraints to the ext_emconf.php
	* Add support to add an XCLASS for alt_doc.php if allready an XCLASS exists.

29-10-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Fixed Bug #3928: Referenced records won't show up in Refindex

22-10-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Fixed Bug #5034: URL length of tab links causes errors in IE7

21-10-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Fixed Bug #3926: Omit Overlay-Records within Linkhandler-DB List
	* Fixed furhter link params in linkhandler link processing

25-09-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Fixed Bug #4479: allow recursive onlyPids - Thanks to Stefano Cecere for providing the patch.

11-08-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Minnor code changes

10-08-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Add some comments
	* Update documentation
	* Minnor code changes

09-08-2009 Daniel Pötzinger
	* New configurations for recordTab:
		noAttributesForm=1
		linkClassName=footnote

07-08-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Fix bug #3784: Tabs in Linkview are jumping
	* Add feature #4183: Add parameter overwriteHandler to the tab configuration
	* Minnor code changes
	* Add feature #4184: Show only the "onlyPids" configured pages in the pagetree

03-08-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Minnor code changes
	* Make the tx_linkhandler_handler::getCurrentRecord more robustness

16-07-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Refactor the tx_linkhandler_handler

16-07-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Minnor code changes
	* Add support for WS preview link generation
	* Add support to view records from different language
	* Increase version number
	* Fix bug #3927: tx_linkhandler_handler should check for the l18n_parent and fix the recordRow

12-07-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Optimize Feature #3414: Use linkhandler for backend record "view" button

11-07-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Fix bug #3867: tslib_cObj::getTypoLink_URL - Return just empty string

10-07-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Fix bug #3675: typolink.returnLast not working with record links?

03-07-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Minor code change

30-06-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Improve added feature #3800
	* Minor code change
	* Update documentation

29-06-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Add Feature #3800: Change the "applyPatch" from ext_conf_template to version_compare in ext_localconf.php

27-06-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Add Feature #3414: Use linkhandler for backend record "view" button

05-06-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Fix bug #3452: link handler should respect blindLinkOptions configuration of TCA
	* Minor code indention

21-04-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* fix Bug #3143
17-04-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* Update Documentation

08-04-2009 Michael Klapper <michael.klapper@aoemedia.de>
	* fix Bug #2626, #3057, #3058, #1488, #1244, #2675
	* Add some PHPDoc
	* Code indention

23-01-2009 Stefano Cecere <stefano.cecere@krur.com>
	* fix Bug #1342 (rtehtmlarea hook name has changed)

07-09-17 Daniel Pötzinger  <poetzinger@aoemedia.de>
	* Initial release

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext: "linkhandler"
#
# Auto generated 20-04-2009 11:43
#
# Manual updates:
# Only the data in the array - anything else is removed by next write.
# "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Generic record link handler',
	'description' => 'Create links to any record (e.g. news)',
	'category' => 'plugin',
	'shy' => 0,
	'version' => '3.0.0',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'stable',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Francois Suter',
	'author_email' => 'typo3@cobweb.ch',
	'author_company' => 'Cobweb Development Sarl',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'typo3' => '7.6.1-7.9.99',
		),
		'conflicts' => array(
			'ch_rterecords',
			'tinymce_rte',
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:23:{s:9:"ChangeLog";s:4:"05f9";s:10:"README.txt";s:4:"ee2d";s:41:"class.tx_linkhandler_browselinkshooks.php";s:4:"c32e";s:32:"class.tx_linkhandler_handler.php";s:4:"d45f";s:21:"ext_conf_template.txt";s:4:"6e68";s:12:"ext_icon.gif";s:4:"f19a";s:17:"ext_localconf.php";s:4:"cc5b";s:14:"ext_tables.php";s:4:"64cd";s:42:"classes/class.tx_linkhandler_recordTab.php";s:4:"3272";s:47:"classes/interface.tx_linkhandler_tabHandler.php";s:4:"1ca3";s:50:"classes/record/class.TBE_browser_recordListRTE.php";s:4:"d7f9";s:51:"classes/record/class.tx_linkhandler_recordsTree.php";s:4:"5148";s:14:"doc/manual.sxw";s:4:"653d";s:19:"doc/wizard_form.dat";s:4:"cfc2";s:20:"doc/wizard_form.html";s:4:"c70c";s:35:"patch/4.1/class.ux_browse_links.php";s:4:"6c5a";s:43:"patch/4.1/class.ux_t3lib_parsehtml_proc.php";s:4:"d0de";s:36:"patch/4.1/class.ux_tslib_content.php";s:4:"c23e";s:50:"patch/4.1/class.ux_tx_rtehtmlarea_browse_links.php";s:4:"4328";s:39:"patch/4.1/class.ux_ux_tslib_content.php";s:4:"df20";s:52:"patch/interfaces/interface.t3lib_browselinkshook.php";s:4:"258f";s:33:"static/link_handler/constants.txt";s:4:"1e06";s:29:"static/link_handler/setup.txt";s:4:"f86a";}',
	'suggests' => array(
	),
);

?>